### ISSUE TITLE
Would just be useful as a library?

### ISSUE THREAD

**Issue by casey** [casey/just#365]
I've hesitated to make `just` available as a library crate, since that would introduce some additional stability guarantees that I don't necessarily want to commit to.

However, it seems entirely possible that it might be useful, so I wanted to open this issue up for discussion. If you can think of ways you might use the parser, lexer, or command runner, please comment on this issue and describe your use case!

Since the `just` crate's version is used to describe the semantic version of the *binary*, I think it would probably be ideal to use a different crate for the library crate, so that it can have an independent version which doesn't move in lockstep with the binary. I am very much not excited about the idea of moving things around to facilitate this though 😂

**Comment by RadicalZephyr** [comment]
I've been toying with the idea of using Just as lightweight server automation accessible from an http endpoint.  This would basically just be a new front-end for invoking just targets with an http request. Something like: `GET /just/build` would be equivalent to `just build` on the command line.

This would really only require an interface that can run a target given a vector of `[target-name, & args]`.  And maybe being able to point to a specific directory where the justfile lives.

Of course, this could also be built by shelling out to Just, but if I'm writing the service in Rust it would seem a shame to add that overhead and deployment complexity.

**Comment by RadicalZephyr** [comment]
Actually, thinking more about it the only API that I would need to be exposed would be a slightly refactored `run` method that took the command line arguments as a `&[&str]` instead of pulling them from the environment directly.

**Comment by casey** [comment]
I really like this idea. Plus, like you said, we wouldn't need to expose and stabilize a large API surface, only command line arguments, which arguably are already subject to stability guarantees.

There would be a few other things that would be needed for this to be useful:
- Currently, `just` cds into the directory containing the justfile. Since the CWD is process-wide, this should be changed so that it instead launches processes in the correct directory and doesn't CD.
- stdout and stderr should be captured and returned to the caller, so they can see what happened.

**Comment by KevinMGranger** [comment]
I like the idea of a library interface. I'm writing a concurrent task runner / viewer, and I'd love for the task specification format to be as simple as a Justfile.

The design is still preliminary, but what it needs so far is to check that certain recipes exist, what kind of arguments they take, and what their dependencies are.

Adding arbitrary metadata to these recipes would be cool too, but that could be extracted from the comment.

Right now, my tool combines the concurrent task running, and creating tasks based off of rust versions and copying your source code. I'd like to separate them, which means the creation part would have to _generate_ a Justfile too. That's easy enough to do with string templating, but a library interface might be cleaner.

**Comment by nim65s** [comment]
Yes, this would be nice. Here is another use case: I currently have multiple commands defined in a justfile, and would like to automate the process of simply launching them all [wrapped in bacon](https://github.com/Canop/bacon/issues/144) in different panels/tabs of my shell. 

I tried to do this in a [zellij plugin](https://zellij.dev/news/new-plugin-system/) to allow dynamic panel/tabs creation and layout depending on the number of `just` commands and the size of the window. But zellij plugin are bundled into wasi, so I can't use something like `std::process::Command::new("just").arg("-l")`, to know what commands to launch. Instead, I could [directly access the justfile](https://zellij.dev/documentation/plugin-api-file-system.html) and use "just-library" to parse it. Alas, `struct Justfile` is `pub(crate)` :(

I'll use a regex to parse this file for now, but I find this a bit sad ^^"

**Comment by willie** [comment]
I've implemented a partial justfile parser in Go because I needed some other tooling bolted on, so a library would be cool, but I don't know how it would work to be useful...

**Comment by willie** [comment]
I'll tell you why a library or modifications to your --dry-run would be useful.

I want to not show secrets (think berglas:// or op://) when it's executing and I need to be able to decode those AFTER that is shown, but before running the command.

The very last command run should be exec instead, so that if you are doing a bunch of stuff before setup, so that you can hand over the process to the last command (a server for example) for debug.

Right now, I have to call --dump to get recipes for completion (I support built-ins and repo level) and then --dry-run for execution, since I have to replace just actually doing the running of the recipes, but it doesn't pass the sigils through and that's hella annoying to try and match the --dump up with the --dry-run output.

I would fork, but I also have a lot of configuration and the secrets decoding that is built in Go already and a fork seems painful to maintain.

So a library would let me do stuff or some flags or a dry-run mode that had a json format output like --dump would make sense.

**Comment by casey** [comment]
@will There's already a library interface, but it's super minimal. See [the docs](https://docs.rs/just/1.47.1/just/). What are the additions you would need for your use-case?

**Comment by will** [comment]
I don’t know. I’ve never used just myself. I’ve seen it scroll by in some nix build logs I think? So as a library hmmmmm. Maybe just to avoid re execing something? I’m just trying to come up with an answer for you. Sorry I’m not more help :(On Mar 17, 2026, at 04:58, Casey Rodarmor ***@***.***> wrote:﻿casey left a comment (casey/just#365)
@will There's already a library interface, but it's super minimal. See the docs. What are the additions you would need for your use-case?

**Comment by casey** [comment]
Looool, sorry @will, wrong person 😅

@willie See my comment above.

**Comment by willie** [comment]
Thanks @casey! To be fair, the entry point is good for inclusion, but hard for deep integration like I referenced in https://github.com/casey/just/issues/365#issuecomment-3917993332. I ended up punting the command runner back out to shell out to individual runners and it's up to users to mask output in their recipes/commands/etc if secrets are involved. I don't get nice integrated completions from the base config chooser (config is the reason my tool exists) but it works. My bias is towards just, but some colleagues like their specific language systems like rake and others. I appreciate the follow up.
